framework/wazuh/core/cluster/hap_helper/proxy.py

```diff
@@ -507,9 +533,44 @@ async def initialize(self):
         await self.api.initialize()
         try:
             (await self.api.get_runtime_info())['version']
+            hard_stop_after = await self.get_hard_stop_after_value()
+            self.hard_stop_after = _convert_to_seconds(hard_stop_after) if hard_stop_after is not None else None
         except (KeyError, IndexError):
             raise WazuhHAPHelperError(3048)
 
+    async def get_hard_stop_after_value(self) -> Optional[str]:
+        """Get the `hard-stop-after` value from the global configuration.
+
+        Returns
+        -------
+        Optional[str]
+            The value of the configuration.
+        """
+        return (await self.api.get_global_configuration()).get('hard_stop_after', None)
+
+    async def set_hard_stop_after_value(self, active_agents: int, chunk_size: int, agent_reconnection_time: int):
+        """Calculate a dynamic value for `hard-stop-after` and set it.
+
+        Parameters
+        ----------
+        active_agents : int
+            Number of active agents.
+        chunk_size : int
+            Max number of agents to be reconnected at once.
+        agent_reconnection_time : int
+            Seconds to sleep after an agent chunk reconnection.
+        """
+        number_of_chunks = active_agents / chunk_size if active_agents > chunk_size else 1
+        hard_stop_after = number_of_chunks * agent_reconnection_time
```

Comment:
As we previously discussed, this formula does not seem to be good enough for calculating the proper `hard_stop_after` value. 

After taking a look at the logs you attached [here](https://github.com/wazuh/wazuh/pull/22578#issuecomment-2022677385), we can see that after adding a new worker (`worker3`), it took **80s** for the helper to complete the migration of all agents. However, the hard_stop_value set was **40s**, so the old haproxy process is killed while the migration is taking place. 

This is because we are not taking into account the sleep time between each Drain/Ready status change for each manager:
1. DRAIN (master, worker1, worker3) -> 5" (hardcoded)
2. RECONNECT/MIGRATE (n_chunks * agent_reconnection_time, 1*10) -> 10"
3. READY (master, worker1, worker3) -> 5" (hardcoded)
4. DRAIN (worker1, worker2, worker3) -> 5" (hardcoded)
5. RECONNECT/MIGRATE (1*10) -> 10"
6. READY (worker1, worker2, worker3) -> 5" (hardcoded)
7. DRAIN (master, worker2, worker3) -> 5" (hardcoded)
8. RECONNECT/MIGRATE (1*10) -> 10"
9. READY (master, worker2, worker3) -> 5" (hardcoded)
10. RECONNECT/EXCEEDING (2*10) -> 20"  

This formula should do the trick, at least in the scenario of the logs you showed, but please, review it and check if there could be other cases where the `hard_stop_after` value still falls short:
```
⌈active_agents / (n_managers * chunk_size)⌉ * n_managers  * agent_reconnection_time + (n_managers * server_admin_state_delay * 2)
```

Reply:
After implementing the new formula I repeated the same test, using `agent_reconnection_chunk_size: 5` and `agent_reconnection_time: 10`.

```bash
2024/04/03 20:07:46 INFO: [HAPHelper] [Main] Proxy was initialized
2024/04/03 20:07:46 DEBUG2: [HAPHelper] [Proxy] Obtained proxy backends
2024/04/03 20:07:46 INFO: [HAPHelper] [Main] Could not find Wazuh backend 'wazuh_cluster'
2024/04/03 20:07:46 DEBUG2: [HAPHelper] [Proxy] Added new proxy backend: 'wazuh_cluster'
2024/04/03 20:07:46 INFO: [HAPHelper] [Main] Added Wazuh backend
2024/04/03 20:07:46 DEBUG2: [HAPHelper] [Proxy] Obtained proxy frontends
2024/04/03 20:07:46 INFO: [HAPHelper] [Main] Could not find Wazuh frontend 'wazuh_cluster_front'
2024/04/03 20:07:47 DEBUG2: [HAPHelper] [Proxy] Added new proxy frontend: 'wazuh_cluster_front'
2024/04/03 20:07:47 INFO: [HAPHelper] [Main] Added Wazuh frontend
2024/04/03 20:07:47 DEBUG: [HAPHelper] [Main] Waiting 12s for workers connections...
2024/04/03 20:07:59 INFO: [HAPHelper] [Main] Setting a value for `hard-stop-after` configuration.
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Receiving parameters {'select': ['node_name', 'version'], 'sort': {'fields': ['version', 'id'], 'order': 'desc'}, 'filters': {'status': 'active'}, 'q': 'id!=000', 'limit': 100000}
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.018s
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Receiving parameters {}
2024/04/03 20:07:59 INFO: [HAPHelper] [Proxy] Set `hard-stop-after` with 30.0 seconds.
2024/04/03 20:07:59 INFO: [HAPHelper] [Main] Starting HAProxy Helper
2024/04/03 20:07:59 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Receiving parameters {}
2024/04/03 20:07:59 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:07:59 INFO: [HAPHelper] [Main] Detected changes in Wazuh cluster nodes. Current cluster: {'master-node': 'wazuh-master', 'worker1': '192.168.128.3', 'worker2': '192.168.128.5'}
2024/04/03 20:07:59 INFO: [HAPHelper] [Main] Attempting to update proxy backend
2024/04/03 20:07:59 DEBUG2: [HAPHelper] [Proxy] Added new server 'master-node' wazuh-master:1514 to backend 'wazuh_cluster'
2024/04/03 20:07:59 DEBUG2: [HAPHelper] [Proxy] Added new server 'worker1' 192.168.128.3:1514 to backend 'wazuh_cluster'
2024/04/03 20:07:59 DEBUG2: [HAPHelper] [Proxy] Added new server 'worker2' 192.168.128.5:1514 to backend 'wazuh_cluster'
2024/04/03 20:07:59 INFO: [HAPHelper] [Main] Setting a value for `hard-stop-after` configuration.
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Receiving parameters {'select': ['node_name', 'version'], 'sort': {'fields': ['version', 'id'], 'order': 'desc'}, 'filters': {'status': 'active'}, 'q': 'id!=000', 'limit': 100000}
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.005s
2024/04/03 20:07:59 DEBUG: [HAPHelper] [D API] Receiving parameters {}
2024/04/03 20:07:59 INFO: [HAPHelper] [Proxy] Set `hard-stop-after` with 32.0 seconds.
2024/04/03 20:07:59 DEBUG: [HAPHelper] [Main] Waiting for new servers to go UP
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Obtained server 'master-node' stats
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker1' stats
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker2' stats
2024/04/03 20:08:00 DEBUG: [HAPHelper] [Main] All new servers are UP
2024/04/03 20:08:00 DEBUG: [HAPHelper] [D API] Receiving parameters {'select': ['node_name', 'version'], 'sort': {'fields': ['version', 'id'], 'order': 'desc'}, 'filters': {'status': 'active'}, 'q': 'id!=000', 'limit': 100000}
2024/04/03 20:08:00 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:08:00 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:08:00 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.005s
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Obtained server 'master-node' stats
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker1' stats
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker2' stats
2024/04/03 20:08:00 DEBUG: [HAPHelper] [Main] Checking for agent balance. Current connections distribution: {'master-node': 2, 'worker1': 2, 'worker2': 1}
2024/04/03 20:08:00 DEBUG: [HAPHelper] [Main] Current balance is under tolerance
2024/04/03 20:08:00 INFO: [HAPHelper] [Main] Migrating 4 connections from server 'master-node'
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker2' to DRAIN state
2024/04/03 20:08:00 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker1' to DRAIN state
2024/04/03 20:08:05 DEBUG: [HAPHelper] [Main] Reconnecting agents
2024/04/03 20:08:05 DEBUG: [HAPHelper] [Main] Agent reconnection chunk size is set to 5. Total iterations: 1
2024/04/03 20:08:05 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['018', '016', '015', '014']}
2024/04/03 20:08:05 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['014', '015', '016', '018']}
2024/04/03 20:08:05 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:08:05 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:08:05 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.010s
2024/04/03 20:08:05 DEBUG: [HAPHelper] [Main] Delay between agent reconnections. Sleeping 10s...
2024/04/03 20:08:15 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker2' to READY state
2024/04/03 20:08:15 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker1' to READY state
2024/04/03 20:08:20 INFO: [HAPHelper] [Main] Waiting for agent connections stability
2024/04/03 20:08:20 DEBUG: [HAPHelper] [Main] Sleeping 60s...
2024/04/03 20:10:21 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:10:21 DEBUG: [HAPHelper] [D API] Receiving parameters {}
2024/04/03 20:10:21 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:10:21 INFO: [HAPHelper] [Main] Detected changes in Wazuh cluster nodes. Current cluster: {'master-node': 'wazuh-master', 'worker1': '192.168.128.3', 'worker2': '192.168.128.5', 'worker3': '192.168.128.4'}
2024/04/03 20:10:21 INFO: [HAPHelper] [Main] Attempting to update proxy backend
2024/04/03 20:10:21 DEBUG2: [HAPHelper] [Proxy] Added new server 'worker3' 192.168.128.4:1514 to backend 'wazuh_cluster'
2024/04/03 20:10:21 INFO: [HAPHelper] [Main] Setting a value for `hard-stop-after` configuration.
2024/04/03 20:10:21 DEBUG: [HAPHelper] [D API] Receiving parameters {'select': ['node_name', 'version'], 'sort': {'fields': ['version', 'id'], 'order': 'desc'}, 'filters': {'status': 'active'}, 'q': 'id!=000', 'limit': 100000}
2024/04/03 20:10:21 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:10:21 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:10:21 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.005s
2024/04/03 20:10:21 DEBUG: [HAPHelper] [D API] Receiving parameters {}
2024/04/03 20:10:21 INFO: [HAPHelper] [Proxy] Set `hard-stop-after` with 80.0 seconds.
2024/04/03 20:10:21 DEBUG: [HAPHelper] [Main] Waiting for new servers to go UP
2024/04/03 20:10:22 DEBUG2: [HAPHelper] [Proxy] Obtained server 'master-node' stats
2024/04/03 20:10:22 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker1' stats
2024/04/03 20:10:22 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker2' stats
2024/04/03 20:10:22 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker3' stats
2024/04/03 20:10:22 DEBUG: [HAPHelper] [Main] Waiting for new servers to go UP
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'master-node' stats
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker1' stats
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker2' stats
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker3' stats
2024/04/03 20:10:23 DEBUG: [HAPHelper] [Main] All new servers are UP
2024/04/03 20:10:23 DEBUG: [HAPHelper] [D API] Receiving parameters {'select': ['node_name', 'version'], 'sort': {'fields': ['version', 'id'], 'order': 'desc'}, 'filters': {'status': 'active'}, 'q': 'id!=000', 'limit': 100000}
2024/04/03 20:10:23 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:10:23 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:10:23 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.005s
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'master-node' stats
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker1' stats
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker2' stats
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained server 'worker3' stats
2024/04/03 20:10:23 DEBUG: [HAPHelper] [Main] Checking for agent balance. Current connections distribution: {'master-node': 7, 'worker1': 7, 'worker2': 6, 'worker3': 0}
2024/04/03 20:10:23 INFO: [HAPHelper] [Main] Migrating 5 connections from server 'worker2'
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker1' to DRAIN state
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'master-node' to DRAIN state
2024/04/03 20:10:23 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker3' to DRAIN state
2024/04/03 20:10:28 DEBUG: [HAPHelper] [Main] Reconnecting agents
2024/04/03 20:10:28 DEBUG: [HAPHelper] [Main] Agent reconnection chunk size is set to 5. Total iterations: 1
2024/04/03 20:10:28 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['013', '012', '011', '003', '002']}
2024/04/03 20:10:28 DEBUG: [HAPHelper] [Main] Delay between agent reconnections. Sleeping 10s...
2024/04/03 20:10:38 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker1' to READY state
2024/04/03 20:10:38 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'master-node' to READY state
2024/04/03 20:10:38 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker3' to READY state
2024/04/03 20:10:43 INFO: [HAPHelper] [Main] Migrating 5 connections from server 'master-node'
2024/04/03 20:10:43 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:10:43 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker2' to DRAIN state
2024/04/03 20:10:43 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker1' to DRAIN state
2024/04/03 20:10:43 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker3' to DRAIN state
2024/04/03 20:10:48 DEBUG: [HAPHelper] [Main] Reconnecting agents
2024/04/03 20:10:48 DEBUG: [HAPHelper] [Main] Agent reconnection chunk size is set to 5. Total iterations: 1
2024/04/03 20:10:48 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['010', '009', '007', '004', '001']}
2024/04/03 20:10:48 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['001', '004', '007', '009', '010']}
2024/04/03 20:10:48 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:10:48 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:10:48 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.012s
2024/04/03 20:10:48 DEBUG: [HAPHelper] [Main] Delay between agent reconnections. Sleeping 10s...
2024/04/03 20:10:58 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker2' to READY state
2024/04/03 20:10:58 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker1' to READY state
2024/04/03 20:10:58 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker3' to READY state
2024/04/03 20:11:03 INFO: [HAPHelper] [Main] Migrating 5 connections from server 'worker1'
2024/04/03 20:11:03 DEBUG2: [HAPHelper] [Proxy] Obtained proxy servers
2024/04/03 20:11:03 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker2' to DRAIN state
2024/04/03 20:11:03 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'master-node' to DRAIN state
2024/04/03 20:11:03 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker3' to DRAIN state
2024/04/03 20:11:08 DEBUG: [HAPHelper] [Main] Reconnecting agents
2024/04/03 20:11:08 DEBUG: [HAPHelper] [Main] Agent reconnection chunk size is set to 5. Total iterations: 1
2024/04/03 20:11:08 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['015', '014', '008', '006', '005']}
2024/04/03 20:11:08 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['005', '006', '008', '014', '015']}
2024/04/03 20:11:08 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:11:08 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:11:08 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.012s
2024/04/03 20:11:08 DEBUG: [HAPHelper] [Main] Delay between agent reconnections. Sleeping 10s...
2024/04/03 20:11:18 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker2' to READY state
2024/04/03 20:11:18 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'master-node' to READY state
2024/04/03 20:11:18 DEBUG2: [HAPHelper] [Proxy] Changed Wazuh server 'worker3' to READY state
2024/04/03 20:11:23 INFO: [HAPHelper] [Main] Balancing exceeding connections after changes on the Wazuh backend
2024/04/03 20:11:23 DEBUG: [HAPHelper] [Main] Reconnecting agents
2024/04/03 20:11:23 DEBUG: [HAPHelper] [Main] Agent reconnection chunk size is set to 5. Total iterations: 1
2024/04/03 20:11:23 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['020', '019', '016', '018', '017']}
2024/04/03 20:11:23 DEBUG: [HAPHelper] [D API] Receiving parameters {'agent_list': ['016', '017', '018', '019', '020']}
2024/04/03 20:11:23 DEBUG: [HAPHelper] [D API] Starting to execute request locally
2024/04/03 20:11:23 DEBUG: [HAPHelper] [D API] Finished executing request locally
2024/04/03 20:11:23 DEBUG: [HAPHelper] [D API] Time calculating request result: 0.011s
2024/04/03 20:11:23 DEBUG: [HAPHelper] [Main] Delay between agent reconnections. Sleeping 10s...
2024/04/03 20:11:33 INFO: [HAPHelper] [Main] Waiting for agent connections stability
2024/04/03 20:11:33 DEBUG: [HAPHelper] [Main] Sleeping 60s...
```

Now we can see three times when the `hard-stop-after` was set.
1. When the helper initiates with three managers and zero agents. The value is `30s`.
2. When the helper created the HAProxy servers, one agent was connected. The value now is `32s` and all the process was made in `21s`.
3. When the worker3 was added and the twenty agents were connected. The value now is `80s` and all the process was made in `72s`